### PR TITLE
Fund policies: GraphQL/display

### PR DESCRIPTION
**Pull request by iherger**
- setup graphql resolvers for fund policies
- display existing fund policies with their respective parameters

Depends on newly implemented `melonjs` functions (now.sh build failing because of that): https://github.com/melonproject/melonjs/pull/40

### CODE REVIEW

File: src/queries/FundPolicies.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+              maxPositions
+            }
+            ... on PriceTolerance {
+              priceTolerance
+            }
+            ... on AssetWhitelist {
+              assetWhitelist
+            }
+            ... on AssetBlacklist {
+              assetWhitelist
```
Comment (fubhy) [inline src/queries/FundPolicies.ts]: That seems wrong.

File: src/components/Routes/Fund/FundRoutes/FundDetails/FundHoldings/FundHoldings.tsx [hunk 2]
```diff
@@ -19,28 +19,31 @@ export const FundHoldings: React.FC<FundHoldingsProps> = ({ address }) => {
...
+        <thead>
+          <S.HeaderRow>
+            <S.HeaderCell>Asset</S.HeaderCell>
+            <S.HeaderCell>Price</S.HeaderCell>
+            <S.HeaderCell>Balance</S.HeaderCell>
+          </S.HeaderRow>
+        </thead>
+        <tbody>
+          {holdings
+            .filter(holding => !isNull(holding))
```
Comment (fubhy) [inline src/components/Routes/Fund/FundRoutes/FundDetails/FundHoldings/FundHoldings.tsx]: I know this was here before but using isNull is not good. Instead, just check if the value is set or not.

File: src/components/Routes/Fund/FundRoutes/FundDetails/FundPolicies/FundPolicies.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import { useFundPoliciesQuery } from '~/queries/FundPolicies';
+import { findToken } from '~/utils/findToken';
+import { useEnvironment } from '~/hooks/useEnvironment';
+
+export interface FundPoliciesProps {
+  address: string;
+}
+
+export const FundPolicies: React.FC<FundPoliciesProps> = ({ address }) => {
+  const environment = useEnvironment();
```
Comment (fubhy) [inline src/components/Routes/Fund/FundRoutes/FundDetails/FundPolicies/FundPolicies.tsx]: By adding an exclamation mark here (you can expect environment to always be set at this point) you can avoid checking for environment further down.

`const environment = useEnvironment()!;`

File: src/components/Routes/Fund/FundRoutes/FundDetails/FundPolicies/FundPolicies.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    return null;
+  }
+
+  return (
+    <>
+      <h1>Policies</h1>
+      {policies.length > 0 && (
+        <S.Table>
+          <tbody>
+            {policies.map(policy => (
```
Comment (fubhy) [inline src/components/Routes/Fund/FundRoutes/FundDetails/FundPolicies/FundPolicies.tsx]: I would use a `switch` statement here:

```
switch (policy.__typename) {
   case 'UserWhitelist': {
     return <UserWhiteListPolicy {...policy} />
   }

  ....
}
```

File: src/components/Routes/Fund/FundRoutes/FundInvest/FundInvest.tsx [hunk 5]
```diff
@@ -1,12 +1,16 @@
 import React from 'react';
-import { Spinner } from '~/components/Common/Spinner/Spinner';
+import { PaddedBody } from '~/components/Common/Styles/Styles';
```
Comment (fubhy) [inline src/components/Routes/Fund/FundRoutes/FundInvest/FundInvest.tsx]: Don't use Styles components directly. Re-export them from local `*.styles.ts` using `* as S from ...`

File: src/components/Routes/Fund/FundRoutes/FundPolicies/FundPolicies.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+export const Policies: React.FC<PoliciesProps> = ({ address }) => {
+  const policyKeys = Object.keys(availablePolicies);
+  const policies = availablePolicies;
+
+  return (
+    <>
+      <PaddedBody>
+        <h1>Risk Profile</h1>
+        <p>Configure the risk management profile of your fund and the rules to be enforced by the smart contracts.</p>
+        <p>&nbsp;</p>
```
Comment (fubhy) [inline src/components/Routes/Fund/FundRoutes/FundPolicies/FundPolicies.tsx]: Avoid implementing spacing with empty characters.

File: src/components/Routes/Fund/FundRoutes/FundPolicies/FundPolicies.tsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+import { PaddedBody } from '~/components/Common/Styles/Styles';
```
Comment (fubhy) [inline src/components/Routes/Fund/FundRoutes/FundPolicies/FundPolicies.tsx]: Same here.

File: src/components/Routes/Fund/FundRoutes/FundPolicies/FundPolicies.tsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import { availablePolicies } from '~/utils/availablePolicies';
+
+export interface PoliciesProps {
+  address: string;
+}
+
+export const Policies: React.FC<PoliciesProps> = ({ address }) => {
+  const policyKeys = Object.keys(availablePolicies);
+  const policies = availablePolicies;
```
Comment (fubhy) [inline src/components/Routes/Fund/FundRoutes/FundPolicies/FundPolicies.tsx]: Redundant.

File: src/components/Routes/Fund/FundRoutes/FundPolicies/FundPolicies.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    <>
+      <PaddedBody>
+        <h1>Risk Profile</h1>
+        <p>Configure the risk management profile of your fund and the rules to be enforced by the smart contracts.</p>
+        <p>&nbsp;</p>
+        <h3>Active policies (0)</h3>
+        <p>No active policies</p>
+        <p>&nbsp;</p>
+
+        <h3>Available policies (6)</h3>
```
Comment (fubhy) [inline src/components/Routes/Fund/FundRoutes/FundPolicies/FundPolicies.tsx]: Use proper counts here instead of hardcoding it (6).

Comment (fubhy) [reply src/components/Routes/Fund/FundRoutes/FundPolicies/FundPolicies.tsx]: I guess this component is not finished yet?

File: src/components/Routes/Fund/FundRoutes/FundRedeem/FundRedeem.tsx [hunk 10]
```diff
@@ -1,12 +1,16 @@
 import React from 'react';
-import { Spinner } from '~/components/Common/Spinner/Spinner';
+import { PaddedBody } from '~/components/Common/Styles/Styles';
```
Comment (fubhy) [inline src/components/Routes/Fund/FundRoutes/FundRedeem/FundRedeem.tsx]: Same here.

File: src/components/Routes/Fund/FundRoutes/FundShutdown/FundShutdown.tsx [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import { PaddedBody } from '~/components/Common/Styles/Styles';
```
Comment (fubhy) [inline src/components/Routes/Fund/FundRoutes/FundShutdown/FundShutdown.tsx]: Same here.

File: src/graphql/schema.graphql [hunk 12]
```diff
@@ -95,6 +95,52 @@ type Vault {
 
 type PolicyManager {
   address: Address!
+  policies: [Policy]
```
Comment (fubhy) [inline src/graphql/schema.graphql]: Make this `[Policy!]!`

File: src/graphql/resolvers/routes/PolicyManager.ts [hunk 13]
```diff
@@ -1,4 +1,10 @@
 import { Resolver } from '~/graphql';
-import { PolicyManager } from '@melonproject/melonjs';
+import { PolicyManager, Policy } from '@melonproject/melonjs';
 
-export const address: Resolver<PolicyManager> = shares => shares.contract.address;
+export const address: Resolver<PolicyManager> = policyManager => policyManager.contract.address;
+
+export const policies: Resolver<PolicyManager> = async (policyManager, _, context) => {
+  const policies = await policyManager.getPolicies();
```
Comment (fubhy) [inline src/graphql/resolvers/routes/PolicyManager.ts]: Don't forget to add the block number from the context when you are making calls.

File: src/graphql/resolvers/policies/PriceTolerance.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+import { Resolver } from '~/graphql';
+import { PriceTolerance } from '@melonproject/melonjs';
+
+export const priceTolerance: Resolver<PriceTolerance> = (policy, _, context) => {
+  const priceToleranceContract = new PriceTolerance(context.environment, policy.contract.address);
+  return priceToleranceContract.getPriceTolerance();
```
Comment (fubhy) [inline src/graphql/resolvers/policies/PriceTolerance.ts]: Don't forget to add the block number from the context when you are making calls.

File: src/graphql/resolvers/policies/Policy.ts [hunk 15]
```diff
@@ -0,0 +1,26 @@
+import { Policy } from '@melonproject/melonjs';
+import { Resolver } from '~/graphql';
+
+export const address: Resolver<Policy> = policy => policy.contract.address;
+
+export const identifier: Resolver<Policy> = policy => policy.getIdentifier();
+
+export const __resolveType: Resolver<Policy> = async (policy, _, __) => {
+  const identifier = await policy.getIdentifier();
```
Comment (fubhy) [inline src/graphql/resolvers/policies/Policy.ts]: Don't forget to add the block number from the context when you are making calls.

File: src/graphql/resolvers/policies/MaxPositions.ts [hunk 16]
```diff
@@ -0,0 +1,7 @@
+import { Resolver } from '~/graphql';
+import { MaxPositions } from '@melonproject/melonjs';
+
+export const maxPositions: Resolver<MaxPositions> = (policy, _, context) => {
+  const priceToleranceContract = new MaxPositions(context.environment, policy.contract.address);
+  return priceToleranceContract.getMaxPositions();
```
Comment (fubhy) [inline src/graphql/resolvers/policies/MaxPositions.ts]: Don't forget to add the block number from the context when you are making calls.

File: src/graphql/resolvers/policies/MaxConcentration.ts [hunk 17]
```diff
@@ -0,0 +1,7 @@
+import { Resolver } from '~/graphql';
+import { MaxConcentration } from '@melonproject/melonjs';
+
+export const maxConcentration: Resolver<MaxConcentration> = (policy, _, context) => {
+  const priceToleranceContract = new MaxConcentration(context.environment, policy.contract.address);
+  return priceToleranceContract.getMaxConcentration();
```
Comment (fubhy) [inline src/graphql/resolvers/policies/MaxConcentration.ts]: Don't forget to add the block number from the context when you are making calls.

File: src/graphql/resolvers/policies/AssetWhitelist.ts [hunk 18]
```diff
@@ -0,0 +1,7 @@
+import { Resolver } from '~/graphql';
+import { AssetWhitelist } from '@melonproject/melonjs';
+
+export const assetWhitelist: Resolver<AssetWhitelist> = (policy, _, context) => {
+  const priceToleranceContract = new AssetWhitelist(context.environment, policy.contract.address);
+  return priceToleranceContract.getMembers();
```
Comment (fubhy) [inline src/graphql/resolvers/policies/AssetWhitelist.ts]: Don't forget to add the block number from the context when you are making calls.

File: src/graphql/resolvers/policies/AssetBlacklist.ts [hunk 19]
```diff
@@ -0,0 +1,7 @@
+import { Resolver } from '~/graphql';
+import { AssetBlacklist } from '@melonproject/melonjs';
+
+export const assetBlacklist: Resolver<AssetBlacklist> = (policy, _, context) => {
+  const priceToleranceContract = new AssetBlacklist(context.environment, policy.contract.address);
+  return priceToleranceContract.getMembers();
```
Comment (fubhy) [inline src/graphql/resolvers/policies/AssetBlacklist.ts]: Don't forget to add the block number from the context when you are making calls.

File: src/components/Routes/Fund/FundRoutes/FundPolicies/FundPolicies.tsx (hunk 9 above)
Comment (iherger) [reply src/components/Routes/Fund/FundRoutes/FundPolicies/FundPolicies.tsx]: Yes, this is work in progress...

**Pull request merged**